src/storage/src/controller/hosts.rs

```diff
@@ -151,15 +151,30 @@ where
         let (host_addr, orchestrated) = match host_config {
             StorageHostConfig::Remote { addr } => (addr, false),
             StorageHostConfig::Managed { allocation, .. } => {
-                (self.start_storage_host(id, allocation).await?, true)
+                (self.ensure_storage_host(id, allocation).await?, true)
             }
         };
-        let existed = self.objects.insert(id, host_addr.clone());
-        assert!(
-            existed.is_none(),
-            "StorageHosts::provision called for already-provisioned ID {id}"
-        );
-        info!("assigned storage object {id} to storage host {host_addr}");
+
+        // True iff we're reprovisioning an existing service address, instead of creating a fresh one
+        let reprovisioning_addr =
+            if let Some(previous_address) = self.objects.insert(id, host_addr.clone()) {
+                if &host_addr == &previous_address {
+                    info!("reprovisioning existing service {} for id {id}", &host_addr);
+                    true
+                } else {
+                    info!(
+                        "reprovisioning service for {id} changed address: {} -> {}",
+                        &previous_address, &host_addr
+                    );
+                    // NB: we could clean up the service here, but better to re-ensure it below.
+                    self.deprovision_host(id, previous_address).await?;
```

Comment:
I think I'm missing something. Where is the storaged process get cleaned up if this call returns true?

Reply:
The short answer is -- in the call to `ensure_storage_host` above, which will make any changes to the set of running hosts required to satisfy our new config, including cleaning up any hosts it replaces. This call cleans the old host out of our internal datastructures. `drop_storage_host(gid)` spins down the kubernetes service for a given gid, and we absolutely don't want that here, since we just reconfigured it!

A lot of the sketchiness in the implementation here is there to deal with "remote" hosts which are actually pointed at a storaged that this process itself is managing. It's my opinion that that's a weird thing to support, but since I'm new and don't really know how folks set this stuff up for testing yet, I left it alone. If you think we could do without supporting that case - I'd love to do a followup clearing all that out, but I'll probably run it by Slack first?

Reply:
It turns out I had to mess with this code anyways to support transitioning from remote to managed and vice versa, so I went for a more direct approach - see top-level comment.